rust/src/sans_io/read.rs

```diff
@@ -0,0 +1,1014 @@
+        self.data.resize(desired_end, 0);
+        self.tail()
+    }
+
+    // resets the RWBuf.
+    fn reset(&mut self) {
+        self.data.clear();
+        self.start = 0;
+        self.end = 0;
+    }
+}
+
+enum ReadingFrom {
+    File,
+    Chunk(ChunkState),
+}
+use ReadingFrom::*;
+
+enum SpanOrRemainder {
+    Span((usize, usize)),
+    Remainder(usize),
+}
+use SpanOrRemainder::*;
+
+/// Options for initializing [`LinearReader`].
+#[derive(Debug, Default, Clone)]
+pub struct LinearReaderOptions {
+    /// If true, the reader will not expect the MCAP magic at the start of the stream.
+    pub skip_start_magic: bool,
+    /// If true, the reader will not expect the MCAP magic at the end of the stream.
+    pub skip_end_magic: bool,
+    /// If true, the reader will yield entire chunk records. Otherwise, the reader will decompress
+    /// and read into chunks, yielding the records inside.
+    pub emit_chunks: bool,
+    // Enables chunk CRC validation. Ignored if `prevalidate_chunk_crcs: true`.
+    pub validate_chunk_crcs: bool,
+    // Enables chunk CRC validation before yielding any messages from the chunk. Implies
+    // `validate_chunk_crcs: true`.
+    pub prevalidate_chunk_crcs: bool,
+    // Enables data section CRC validation.
+    pub validate_data_section_crc: bool,
+}
+
+impl LinearReaderOptions {
+    pub fn with_skip_start_magic(self, skip_start_magic: bool) -> Self {
+        Self {
+            skip_start_magic,
+            ..self
+        }
+    }
+    pub fn with_skip_end_magic(self, skip_end_magic: bool) -> Self {
+        Self {
+            skip_end_magic,
+            ..self
+        }
+    }
+    pub fn with_emit_chunks(self, emit_chunks: bool) -> Self {
+        Self {
+            emit_chunks,
+            ..self
+        }
+    }
+    pub fn with_validate_chunk_crcs(self, validate_chunk_crcs: bool) -> Self {
+        Self {
+            validate_chunk_crcs,
+            ..self
+        }
+    }
+    pub fn with_prevalidate_chunk_crcs(self, prevalidate_chunk_crcs: bool) -> Self {
+        Self {
+            prevalidate_chunk_crcs,
+            ..self
+        }
+    }
+    pub fn with_validate_data_section_crc(self, validate_data_section_crc: bool) -> Self {
+        Self {
+            validate_data_section_crc,
+            ..self
+        }
+    }
+}
+
+/// A mutable view that allows the user to write new MCAP data into the [`LinearReader`].
+///
+/// The user is expected to copy up to `self.buf.len()` bytes into `self.buf`, then call
+/// `set_filled(usize)` to notify the reader of how many bytes were successfully read.
+pub struct InputBuf<'a> {
+    pub buf: &'a mut [u8],
+    last_write: &'a mut Option<usize>,
+}
+
+impl<'a> InputBuf<'a> {
+    /// Notify the reader that `written` new bytes are available. Only call this method after
+    /// copying data into [`self.buf`].
+    pub fn set_filled(&'a mut self, written: usize) {
+        *self.last_write = Some(written);
```

Comment:
What makes you think that? I originally had that behavior (where calling set_filled twice would add the two written values together) but felt it was unintuitive for a function called `set_*`. I tried making this clear in the docstring as well.

Reply:
I think it was partly because of the wording of the comment:

> Notify the reader that `written` new bytes are available.

"new bytes" sounds like "new bytes since the last time this function was called".